### PR TITLE
cluster: copy active cluster hosts during warming

### LINKED ISSUE #5168: load balancer stat inconsistency
While investigating some 5xx issue in a cluster found the following stats for the cluster that has problems.
```
cluster.test_cluster.upstream_cx_none_healthy 10
cluster.test_cluster.lb_healthy_panic 10
cluster.test_cluster.membership_total : 3
cluster.test_cluster.membership_healthy: 3
```
While membership metrics show all the members as healthy, the `cx_none_healthy` and `lb_healthy_panic` incrementing seems to be inconsistent. Is this a bug or is there a scenario where this can happen?

### PR DESCRIPTION
**Pull request by ramaraochavali**
*Description*: Copies Active Cluster hosts during warming complete phase, if warming complete is triggered by empty update.
*Risk Level*: Medium
*Testing*: Added a test case to cover this case.
*Docs Changes*: N/A
*Release Notes*: N/A
Fixes #5168

### CODE REVIEW

File: source/common/upstream/cluster_manager_impl.cc [hunk 1]
```diff
@@ -477,6 +477,37 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
...
+          const auto& host_sets =
+              active_cluster_entry.cluster_->prioritySet().hostSetsPerPriority();
+          for (size_t priority = 0; priority < host_sets.size(); ++priority) {
+            const auto& host_set = host_sets[priority];
+            HostVectorConstSharedPtr hosts_copy(new HostVector(host_set->hosts()));
+            HostsPerLocalityConstSharedPtr hosts_per_locality_copy =
+                host_set->hostsPerLocality().clone();
+            cluster_entry.cluster_->prioritySet().updateHosts(
+                priority, HostSetImpl::partitionHosts(hosts_copy, hosts_per_locality_copy),
+                host_set->localityWeights(), {}, {}, absl::nullopt);
```
Comment (ramaraochavali) [inline source/common/upstream/cluster_manager_impl.cc]: I have passed `absl::nullopt` for over provisioning factor with the assumption that, it should use the same over provisioning factor that active cluster hosts are built with. If over provisioning factor is changed, we should expect the management server to send EDS updates? or should I pass what ever is there is warming cluster config?

**Comment by ramaraochavali** [comment]
@htuch @snowp PTAL.

**Comment by htuch** [comment]
@snowp for first pass, I figure Snow is our resident expert here.

File: source/common/upstream/cluster_manager_impl.cc [hunk 2]
```diff
@@ -479,6 +479,37 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
       // Otherwise, applyUpdates() will fire with a dangling cluster reference.
       updates_map_.erase(cluster_name);
 
+      // If the active cluster has priority sets and the warming cluster does not have them, it
```
Comment (snowp) [inline source/common/upstream/cluster_manager_impl.cc]: There's only one priority set per cluster, so this wording is a bit strange

File: source/common/upstream/cluster_manager_impl.cc [hunk 3]
```diff
@@ -479,6 +479,37 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
...
+      // means that onConfigUpdate is triggered for EDS type of cluster is triggered with no
```
Comment (snowp) [inline source/common/upstream/cluster_manager_impl.cc]: Maybe `means that onConfigUpdate was triggered by an EDS update that had no references to this cluster`?

File: source/common/upstream/cluster_manager_impl.cc [hunk 4]
```diff
@@ -479,6 +479,37 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
...
+      // reference to this cluster. In such cases, copy the active cluster priority set to the
+      // warming cluster otherwise the active cluster hosts will be cleared after warming is
```
Comment (snowp) [inline source/common/upstream/cluster_manager_impl.cc]: nit: I'd rephrase this to say something like `to prevent the hosts from being cleared after warming`

File: source/common/upstream/cluster_manager_impl.cc [hunk 5]
```diff
@@ -479,6 +479,37 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
...
+      // completed. See https://github.com/envoyproxy/envoy/issues/5168 for more context.
+      const auto active_it = active_clusters_.find(cluster_name);
+      if (active_it != active_clusters_.end()) {
+        auto& active_cluster_entry = *active_it->second;
```
Comment (snowp) [inline source/common/upstream/cluster_manager_impl.cc]: nit: I think this can be const?

File: source/common/upstream/cluster_manager_impl.cc [hunk 6]
```diff
@@ -479,6 +479,37 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
...
+        if (active_cluster_entry.cluster_->prioritySet().hostSetsPerPriority().size() != 0 &&
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: >  is it just a roundabout way of checking whether the active cluster is empty and that the warming one is not?

It is the other way around. It checks if active cluster has some hosts but not the warming cluster. I think it makes sense to add `empty`  method to `PrioritySet`. Let me try that.

> Also, how does this check work with the fact that we always create a P=0 `HostSet` in the `ClusterImplBase` ctor? Are there actually any scenarios in which we'll hit this code and the active cluster has no `HostSet`s?

Yes. For active clusters it does not copy any thing if there are no hosts.

File: source/common/upstream/cluster_manager_impl.cc (hunk 2 above)
Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: Yeah, I meant hosts - will change

File: source/common/upstream/cluster_manager_impl.cc [hunk 7]
```diff
@@ -479,6 +479,37 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
...
+          ENVOY_LOG(
+              debug,
+              "copying host set from active cluster {} as warming cluster does not have host set",
+              cluster_name);
+
+          const auto& host_sets =
+              active_cluster_entry.cluster_->prioritySet().hostSetsPerPriority();
+          for (size_t priority = 0; priority < host_sets.size(); ++priority) {
+            const auto& host_set = host_sets[priority];
+            HostVectorConstSharedPtr hosts_copy(new HostVector(host_set->hosts()));
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: I tried, but did not know to do it - I think it may be related to this comment https://github.com/envoyproxy/envoy/blob/1c9d6d4359aa4e1a1d0e791b388e89e6bb3983b1/source/common/upstream/cluster_manager_impl.cc#L655. LMK if you have better idea on how to do it?

File: source/common/upstream/cluster_manager_impl.cc [hunk 8]
```diff
@@ -479,6 +479,37 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
...
+            HostsPerLocalityConstSharedPtr hosts_per_locality_copy =
+                host_set->hostsPerLocality().clone();
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: same reason as above.

File: source/common/upstream/cluster_manager_impl.cc (hunk 1 above)
Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: The `active_host_set` has the `overProvisioningFactor` preserved.  We can pass it. I will change it like that.

File: test/common/upstream/cluster_manager_impl_test.cc [hunk 9]
```diff
@@ -1209,6 +1209,83 @@ TEST_F(ClusterManagerImplTest, DynamicAddRemove) {
...
+  EXPECT_TRUE(cluster_manager_->addOrUpdateCluster(update_cluster, "", dummyWarmingCb));
+  checkStats(1 /*added*/, 1 /*modified*/, 0 /*removed*/, 1 /*active*/, 1 /*warming*/);
+  cluster2->initialize_callback_();
+
+  checkStats(1 /*added*/, 1 /*modified*/, 0 /*removed*/, 1 /*active*/, 0 /*warming*/);
+
+  EXPECT_EQ(cluster2->info_, cluster_manager_->get("fake_cluster")->info());
+  EXPECT_EQ(1UL, cluster_manager_->clusters().size());
+
+  // Validate that the host updates are pushed to tls clusters.
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply test/common/upstream/cluster_manager_impl_test.cc]: I was not sure what else state is there on TLS that we should be passed. Agee that hosts would be there. But I could not find a way to check and control `postThreadLocalClusterUpdate` call? LMK if there is an easy way to do it - Since this is rare case, may be it is OK To push TLS state if it is complicated to control?

File: source/common/upstream/cluster_manager_impl.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (snowp) [reply source/common/upstream/cluster_manager_impl.cc]: ok, thanks for explaining

File: test/common/upstream/cluster_manager_impl_test.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (snowp) [reply test/common/upstream/cluster_manager_impl_test.cc]: If you use the `MockedUpdatedClusterManagerImpl` defined in cluster_manager_impl_test then you can add `EXPECT_CALL` on `local_cluster_update_` to verify when the TLS posting happens. 

I think first step is to verify wether we're seeing an additional TLS update by checking it in test, then we can discuss separately whether we want to add additional logic to prevent duplicate updates (if that is even happening).

File: source/common/upstream/cluster_manager_impl.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (snowp) [reply source/common/upstream/cluster_manager_impl.cc]: Ah, the issue is that `partitionHosts` takes a vector of const pointers, while hosts() returns non-const pointers. In that case, mind adding a TODO here as well? I think it would make sense to look into optimizing away these copies at a later time.

File: test/common/upstream/cluster_manager_impl_test.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply test/common/upstream/cluster_manager_impl_test.cc]: I am sure that is happening because there is no checks to prevent that, I will add a test to confirm the same.

File: source/common/upstream/cluster_manager_impl.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: added the TODO

File: test/common/upstream/cluster_manager_impl_test.cc [hunk 10]
```diff
@@ -1209,6 +1209,154 @@ TEST_F(ClusterManagerImplTest, DynamicAddRemove) {
...
+  // Now trigger warming of this cluster with no hosts.
+  auto update_cluster = defaultStaticCluster("fake_cluster");
+  update_cluster.mutable_per_connection_buffer_limit_bytes()->set_value(12345);
+
+  std::shared_ptr<MockClusterRealPrioritySet> cluster2(new NiceMock<MockClusterRealPrioritySet>());
+  EXPECT_CALL(factory_, clusterFromProto_(_, _, _, _)).WillOnce(Return(cluster2));
+  EXPECT_CALL(*cluster2, initializePhase()).Times(0);
+
+  // Validate that TLS updates are triggered correctly after warming.
+  EXPECT_CALL(local_cluster_update_, post(_, _, _))
```
Comment (snowp) [inline test/common/upstream/cluster_manager_impl_test.cc]: This actually tells me that we're *not* triggering a TLS update when we call `updateHosts` in this PR, since we're only seeing this call once. This call is coming from when we explicitly call `postThreadLocalClusterUpdate` in `onClusterInit`. I believe we're not incurring the extra call because the call to `updateHosts` comes *before* we've wired up the callback that triggers TLS updates.

Comment (ramaraochavali) [reply test/common/upstream/cluster_manager_impl_test.cc]: Yes. That is correct. updateHosts does not trigger a TLS update. Do you think we are good with this PR ?

Comment (snowp) [reply test/common/upstream/cluster_manager_impl_test.cc]: Yeah, thanks for taking the time to verify this behavior.

File: source/common/upstream/upstream_impl.h [hunk 11]
```diff
@@ -413,6 +413,15 @@ class PrioritySetImpl : public PrioritySet {
                    const HostVector& hosts_removed,
                    absl::optional<uint32_t> overprovisioning_factor = absl::nullopt) override;
 
+  bool empty() const override {
+    for (auto const& host_set : host_sets_) {
+      if (host_set->hosts().size() > 0) {
```
Comment (snowp) [inline source/common/upstream/upstream_impl.h]: nit: use `empty()`

File: source/common/upstream/cluster_manager_impl.cc [hunk 12]
```diff
@@ -473,18 +473,46 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
...
     cluster_entry->cluster_->initialize([this, cluster_name, cluster_warming_cb] {
       auto warming_it = warming_clusters_.find(cluster_name);
-      auto& cluster_entry = *warming_it->second;
+      auto& warming_cluster_entry = *warming_it->second;
 
       // If the cluster is being updated, we need to cancel any pending merged updates.
       // Otherwise, applyUpdates() will fire with a dangling cluster reference.
       updates_map_.erase(cluster_name);
 
+      // If the active cluster has hosts in priority set and the warming cluster does not have them,
```
Comment (mattklein123) [inline source/common/upstream/cluster_manager_impl.cc]: I didn't follow this conversation, so this was already probably discussed, but isn't this dangerous? Isn't there a case in which this can lead to stale hosts that the management server legitimately does not want the Envoy to have? (Something about this seems off to me).

Comment (snowp) [reply source/common/upstream/cluster_manager_impl.cc]: Yeah I think you're right: There's nothing preventing us from doing this copying for a legitimate zero host EDS update following the CDS update, so I think we're missing a piece here before we can merge this. WDYT @ramaraochavali?

Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: That is a fair point. My intention was to handle this case
https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/eds.cc#L48 - where onConfigUpdate was called with out this cluster reference.  I thought I read it in the xds_protocol that when empty hosts are sent we should work with existing data. Reading it again "When a requested resource is missing in a RDS or EDS update, Envoy will retain the last known value for this resource." - I misinterpreted this to empty hosts, it is essentially talking about the case I am trying to fix. 

I will add an additional condition to trigger this only when `onConfigUpdate` is called with the above condition.

**Comment by ramaraochavali** [comment]
@mattklein123 @snowp handled the case of zero host EDS update and added a test to verify the behaviour as well. Should be good now, PTAL.

File: source/common/upstream/cluster_manager_impl.cc [hunk 13]
```diff
@@ -473,18 +473,45 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
...
     cluster_entry->cluster_->initialize([this, cluster_name, cluster_warming_cb] {
       auto warming_it = warming_clusters_.find(cluster_name);
-      auto& cluster_entry = *warming_it->second;
+      auto& warming_cluster_entry = *warming_it->second;
 
       // If the cluster is being updated, we need to cancel any pending merged updates.
       // Otherwise, applyUpdates() will fire with a dangling cluster reference.
       updates_map_.erase(cluster_name);
 
+      // If onConfigUpdate was triggered by an EDS update that had no references to this
```
Comment (mattklein123) [inline source/common/upstream/cluster_manager_impl.cc]: I think I'm still missing something here. Why isn't the management server responding with the hosts that the Envoy should have? Can you potentially add more comments or we can discuss in this comment thread until we are all on the same page?

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc]: In reading through https://github.com/envoyproxy/envoy/issues/5168 I'm wondering why this isn't a management server issue. Can you just not have your management server enter the LB pool until it's fully ready to serve requests?

Comment (htuch) [reply source/common/upstream/cluster_manager_impl.cc]: I think the best way to think this through is with a concrete example. Let's imaging we're switching from port 80 to 443 (EDS) and HTTP to HTTPS to backend (CDS).

@mattklein123 are you concerned that we _never_ see the 443 EDS host update? Or that we temporarily won't see it during config update?

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc]: An example would be great. My concern here is that it's not clear to me that this is an Envoy issue vs. a management server issue. I.e., why is the management sever accepting connections if it can't actually serve them? Is there a chance that this workaround is going to change legitimate behavior such as the management server wanting to remove all hosts when a cluster has been refreshed, etc. 

I think I would mainly just like to better understand why we think this is the right fix to do here.

Comment (htuch) [reply source/common/upstream/cluster_manager_impl.cc]: My thinking is that the management server always has the capability to send an empty host update (or the 443 hosts in the above example). There may be a period of inconsistency, but arguably that's inevitable when not doing make-before-break consistency.

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc]: I think I'm probably missing something fundamental, so apologies, but why would the the management server send an empty update if there are hosts?

Comment (htuch) [reply source/common/upstream/cluster_manager_impl.cc]: This is the example you gave in:

> Is there a chance that this workaround is going to change legitimate behavior such as the management server wanting to remove all hosts when a cluster has been refreshed, etc.

Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: @mattklein123  just to be clear, management server is not sending empty update for this cluster if there are hosts. It sent ClusterLoadAssignment response for some other cluster (there is no reference to this cluster in the response) and we call `onConfigUpdate` for this cluster from here with empty because we did not find any resources for this cluster https://github.com/envoyproxy/envoy/blob/master/source/common/config/grpc_mux_impl.cc#L171
Normally this is fine because if it is initialized we just increment the stat and onPreInit would do nothing. But if it is in the initializing(warming) state, it is clearing the clusters.

File: source/common/upstream/health_discovery_service.h [hunk 14]
```diff
@@ -55,6 +55,7 @@ class HdsCluster : public Cluster, Logger::Loggable<Logger::Id::upstream> {
   Outlier::Detector* outlierDetector() override { return outlier_detector_.get(); }
   const Outlier::Detector* outlierDetector() const override { return outlier_detector_.get(); }
   void initialize(std::function<void()> callback) override;
+  bool isBeingInitializedByEmptyConfigUpdate() const override { return false; }
```
[the remark this answers is not in the document]
Comment (ramaraochavali) [reply source/common/upstream/health_discovery_service.h]: makes sense. Will change

File: source/common/upstream/cluster_manager_impl.cc (hunk 13 above)
Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: Added more docs. See if that makes sense.

File: source/common/upstream/cluster_manager_impl.cc [hunk 15]
```diff
@@ -473,18 +473,52 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& clust
...
     cluster_entry->cluster_->initialize([this, cluster_name, cluster_warming_cb] {
       auto warming_it = warming_clusters_.find(cluster_name);
-      auto& cluster_entry = *warming_it->second;
+      auto& warming_cluster_entry = *warming_it->second;
 
       // If the cluster is being updated, we need to cancel any pending merged updates.
       // Otherwise, applyUpdates() will fire with a dangling cluster reference.
       updates_map_.erase(cluster_name);
 
+      // If management server sends a EDS response, for any other cluster grpc_mux_impl calls
```
Comment (mattklein123) [inline source/common/upstream/cluster_manager_impl.cc]: @ramaraochavali @snowp @htuch you all are clearly fine with this, so I'm not understanding something, but why does an EDS response for any other cluster cause an empty update for this cluster? Isn't that a bug?

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc]: OK I think I understand, we are looking at this text in the XDS protocol docs: "When a requested resource is missing in a RDS or EDS update, Envoy will retain the last known value for this resource." Right?

I guess this seems OK, though I do wonder if this clause is adding extra complexity that we don't really need, especially with incremental coming. Meaning, it seems like your management server could not accept new connections until it's ready to serve them and we can avoid all of this extra code, potential for bugs, behavior differences, etc. That would be my preference.

If we do decide to keep this, can we spell out the clause we are relying on here in the docs and link to them?

Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: @htuch knows better but IIRC it is required for initial loading otherwise Envoy was struck at initialization - I tried to change that behaviour at some point via https://github.com/envoyproxy/envoy/pull/4276 but got a regression https://github.com/envoyproxy/envoy/issues/4485 and reverted it via https://github.com/envoyproxy/envoy/pull/4490

Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: > we are looking at this text in the XDS protocol docs: "When a requested resource is missing in a RDS or EDS update, Envoy will retain the last known value for this resource." Right?

Yes.

> If we do decide to keep this, can we spell out the clause we are relying on here in the docs and link to them?

I can add the clause in the doc, but we need this fix on Envoy side - it should not clear hosts when it gets EDS response for some other cluster. For EDS/RDS we are not mandating to send all responses every time (Which is the right thing to do). Things may change with incremental but we can deal with it when it comes.

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc]: @ramaraochavali we just discussed this issue briefly on the community call. I think we might need to have a meeting on this. Given that chat, I'm still pretty uncomfortable with this change for 2 reasons:
1) It's not clear to me that's correct to copy the hosts from an old cluster to a new cluster.
2) It adds a bunch of complexity that might lead to hard to understand edge cases.
3) It really seems to me like you could fix this in your management server: don't accept connections to a management server (do not report healthy to your LB) until the management server can fully serve connections.

I know @htuch as some thoughts here so will let him weigh in and then we can maybe do a meeting if needed?

Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: @mattklein123 Sure. Would be happy to chat if needed on this. 

Some clarifications.

> It really seems to me like you could fix this in your management server

IMO, that does not solve the problem completely (btw, we are on our way to do that already) since most of the discovery backends like Zookeeper are event driven. It is difficult for the management server to ensure that it has all the discovery updates for all the clusters given that clusters are also dynamic in nature before it claims it self ready. Let us assume a case which is failing in our deployments - There are two clusters - cluster x and cluster y that are in warming state when an Envoy connected to a management server instance. Envoy sends EDS requests for x and y. When the management server gets update from its discovery backend for x - we will send back the `ClusterLoadAssignment` for cluster x and that response does not have reference to cluster y. Because of the way Envoy works today, on this EDS response both x and y finish warming - x with the hosts sent, y with empty hosts. Later say after 5 mins  as an example our discovery backend sends notification of y hosts and we update it with another EDS response to Envoy. Envoy now updates the hosts correctly. During this period, the requests for cluster y would fail - I think the fundamental problem here is that Envoy is finishing warming of cluster y when management server did not send any EDS response for it - which should be fixed?

Regarding host copy approach, that originated from @htuch 's idea here https://github.com/envoyproxy/envoy/issues/5168#issuecomment-460296843.  Since this copy is just limited to this specific case, it may be OK.

If there is a semantic concern on hosts copy and may lead to edge case issues, we can think about alternatives. Couple of alternatives that come to my mind are 

1. change grpc_mux_impl functionality here https://github.com/envoyproxy/envoy/blob/master/source/common/config/grpc_mux_impl.cc#L171 to call onConfigUpdate only if there are resources. But as I mentioned as mentioned above it had some side effects during Envoy startup and we had to revert it.
2. If the cluster is in warming state, do not finish it's initialization - so basically do not call `onPreInitComplete` here https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/eds.cc#L50 on empty update. So this ensure cluster warming would only be complete when EDS response for that cluster comes. Not sure if there will be any side effects to this. @htuch may know better.

Comment (htuch) [reply source/common/upstream/cluster_manager_impl.cc]: So, I think we should clarify whether this is about update warming only or initial warming. I thought it was the former, but in the above comment from @ramaraochavali, this seems like it applies to initial warming as well.

In general, I think the main gap I see in a common understanding here is what does "warming" mean for clusters or listeners that are updating? What are the guarantees we provide when update warming is completed? Do we demand that the management server send an updated RDS/EDS or not? If we can answer this, then other issues become simpler.

Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: @htuch you are right. our problem of hosts getting cleared only happens for updating warming case. Initial warming since there does not exist any hosts, we do not see any problem. Sorry if my above comment created that confusion.

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc]: > It is difficult for the management server to ensure that it has all the discovery updates for all the clusters given that clusters are also dynamic in nature before it claims it self ready.

I see what you are saying, and I agree we have some problems to fix here, but this still doesn't make sense to me. What do you do if a new Envoy comes up and connects to management server that doesn't have a complete knowledge of the topology? How would that work? Doesn't your management server need to initialize with the most recent data and then merge in updates and send them out as they are available?

Of your proposed solutions, my vote is to focus on either (1) or (2), that seems much more correct to me than what we are proposing here. @htuch @snowp thoughts?

Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: > What do you do if a new Envoy comes up and connects to management server that doesn't have a complete knowledge of the topology? How would that work? Doesn't your management server need to initialize with the most recent data and then merge in updates and send them out as they are available?

With the readiness check in our management we ensure the management server is upto date with the latest cluster information and latest cluster information is sent to newly connected Envoys. As far as EDS updates go, we send whatever latest information we have in the management server at the point in time when the Envoy connected - rest would go in the async mode as and when they are available. With retries in place we do not see a big problem with this.

> Of your proposed solutions, my vote is to focus on either (1) or (2), that seems much more correct to me than what we are proposing here

If we do not go with what is here in this PR, my vote would be for us to try (2) because as I mentioned earlier (1) has been tried for some other reason has been found to have some regressions. However I think the question that @htuch raised above still need to be answered so that we go in the direction that is aligned with future. 
Also slight addition to his question of "
Do we demand that the management server send an updated RDS/EDS or not?" - "Do we demand that the management server send an updated RDS/EDS or not in a **single** response"  because that is what is causing problem here?

Based on what we decide here I will change the direction of this PR.

Comment (snowp) [reply source/common/upstream/cluster_manager_impl.cc]: I agree with @htuch that this boils down to how we want cluster warming to work wrt EDS updates. If we want cluster warming to just rebuild the cluster and continue to use the existing endpoints, then something similar to what we're doing in this PR seems necessary. 

From the beginning I was for the idea of delaying cluster warming until an EDS response came back, but @htuch brought up some good points on the make before break policy in the issue so now I'm less certain about what the right thing to do is.

To me this seems like primarily a management server concern, but I can understand the desire for less strict management servers like what @ramaraochavali is describing. FWIW We're using Zookeeper for service discovery data on our end and we've had no issues with this, presumably due to how we generate both CDS and EDS from the same ZK data. 

If we are going to require an EDS response it seems kinda strange to require it only on cluster warming and not cluster init, so (1) seems like right thing to me, but from the regression that caused it seems like people are expecting to be able to initialize from an empty EDS response, so that'd be a breaking change. We might be stuck with (2).

Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: > FWIW We're using Zookeeper for service discovery data on our end and we've had no issues with this, presumably due to how we generate both CDS and EDS from the same ZK data.

That is one and another point to consider here is this happens only when cluster property change (and hence goes in to warming mode) and the control plane is going through rolling upgrades. In other cases we are also fine.

Another thing longer term to consider is what cluster change need warming i.e. EDS updates for example I change some timeout or some other unrelated setting - why do we have to go through warmup cycle (I understand for simplicity we are doing that by just computing the hash on config)? or for cluster update case can we just skip warming and put the onus management server to send EDS updates if the cluster change needs updated end points?

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc]: > If we want cluster warming to just rebuild the cluster and continue to use the existing endpoints, then something similar to what we're doing in this PR seems necessary

IMO this is not what we want. If a cluster is being warmed it is either new or different (as defined by config) and as such I think it needs to start empty and get hosts via an EDS fetch.

> If we are going to require an EDS response it seems kinda strange to require it only on cluster warming and not cluster init, so (1) seems like right thing to me, but from the regression that caused it seems like people are expecting to be able to initialize from an empty EDS response, so that'd be a breaking change. We might be stuck with (2)

I agree, and IMO we should require a named response before finishing init. I think this might be less painful if https://github.com/envoyproxy/envoy/issues/5862 is implemented (initial fetch timeout to release a cluster if the management server does not respond).

Comment (htuch) [reply source/common/upstream/cluster_manager_impl.cc]: >  IMO this is not what we want. If a cluster is being warmed it is either new or different (as defined by config) and as such I think it needs to start empty and get hosts via an EDS fetch.

This all boils down to (im)mutability. It seems hard to guarantee consistency in the presence of mutability of clusters, since, you might in the case of CDS update, still have a stale EDS on the server. So, even if you do wait for an EDS update, you might not get anything meaningful in response. So, if we delay warming and force an EDS update for only a weak improvement in the consistency of mutable CDS clusters.

The management server can still do an EDS update immediately after CDS, even if we reuse existing EDS resources, we just get a period of potential inconsistency.

I think it comes down to whether a management server can guarantee whether an EDS update following a CDS update truly reflect the new CDS state. In some setups this is true, but in others it isn't. 

A more robust pattern would be to force folks to make all resources immutable. I know this is a broader API design question, but it sure would simplify.

Comment (htuch) [reply source/common/upstream/cluster_manager_impl.cc]: Maybe one way to simplify this discussion is to come up with a bunch of user stories, e.g. "As an Envoy user, I want to upgrade a service from HTTP to HTTPS.. the recommended steps are ...".

Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: > The management server can still do an EDS update immediately after CDS, even if we reuse existing EDS resources, we just get a period of potential inconsistency.

I think it comes down to whether a management server can guarantee whether an EDS update following a CDS update truly reflect the new CDS state. In some setups this is true, but in others it isn't.

I agree with this and it seems potential inconsistency can not be avoided in all cases.

I think we need to think about a short term solution and longer term solutions like immutable resources etc

For short term we have couple of solutions
- Reuse existing EDS resources till management server provides a new set of resources matching the CDS update (Whether this CDS update requires a new set of EDS resources is another question - which we are assuming right now that all CDS updates require EDS response which is kind  of looks not correct)
- Do not finish warming till a named response comes for updated warming case

Do we have any other ideas?

Longer term, for sure we should come up with bunch of user stories and think about solutions like immutable resources etc as @htuch  mentioned.

Comment (mattklein123) [reply source/common/upstream/cluster_manager_impl.cc]: @ramaraochavali take a look at a convo I had with @htuch in Envoy slack: https://envoyproxy.slack.com/archives/CEFDKQ3RQ/p1551391835017700

I'm pretty strongly of the opinion that we should not finish init/warming until we explicitly receive a named response for an EDS fetch. I think this is the most clear solution and IMO the behavior that most people would expect. 

@htuch brings up the valid point that this may not work in all cases if the hosts have changed, but IMO this is very rare, and we should offer guidance in the XDS docs that this edge case should be handled by a cluster rename and correct sequencing.

Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: @mattklein123 @htuch @snowp  Created this doc https://docs.google.com/document/d/1Ca4YX9XNnV9rjTR7U0_xyCxUDKfoSun8pWSUNO9N-tY/edit?usp=sharing so that we can discuss there.

Comment (ramaraochavali) [reply source/common/upstream/cluster_manager_impl.cc]: OK. So since we got consensus on finishing warming only on named responses, I am going to close this PR and open another PR with a fix.
